File: app/Sources/Settings/ConfigSync.swift
import Foundation
import CryptoKit

/// Automatic sync glue over `SyncStore` + `ConfigService`: push the encrypted
/// bundle to the user's folder when the app backgrounds, and pull + merge it on
/// launch. The provider only ever sees ciphertext (no account, no server of
/// ours). Additive merge — matches the manual import. Because the folder +
/// passphrase were configured by the user, auto-pull applies every category
/// (hosts, trusted keys, settings, software keys) without the review dialog.
@MainActor
enum ConfigSync {
    /// Push the current config (incl. software keys) if auto-sync is on and it
    /// changed since the last sync. Safe to call often.
    static func autoPush(sync: SyncStore, hosts: HostStore, knownHosts: KnownHostsStore, agent: AgentStore, snippets: SnippetStore) {
        let start = Date()
        DiagLog.log("sync", "autoPush start configured=\(sync.isConfigured) passphrase=\(sync.hasPassphrase)")
        guard sync.autoEnabled else {
            DiagLog.log("sync", "autoPush skip disabled")
            return
        }
        guard let pass = sync.loadPassphrase() else {
            DiagLog.log("sync", "autoPush skip no passphrase")
            return
        }
        do {
            let softwareKeys = agent.exportableKeys().count
            let securityKeys = agent.exportableSecurityKeyMetadata().count
            DiagLog.log("sync", "autoPush export begin hosts=\(hosts.hosts.count) knownHosts=\(knownHosts.knownHosts.allEntries.count) softwareKeys=\(softwareKeys) securityKeys=\(securityKeys) snippets=\(snippets.snippets.count)")
            let data = try ConfigService.export(hosts: hosts, knownHosts: knownHosts, agent: agent,
                                                snippets: snippets, includeKeys: true, passphrase: pass)
            // Hash the plaintext snapshot (not the ciphertext — encryption is salted
            // and differs each time) so an unchanged config doesn't rewrite the file.
            let plain = try ConfigService.export(hosts: hosts, knownHosts: knownHosts, agent: agent,
                                                 snippets: snippets, includeKeys: true, passphrase: nil)
            let hash = SHA256.hash(data: plain).map { String(format: "%02x", $0) }.joined()
            guard hash != sync.lastHash else {
                DiagLog.log("sync", "autoPush skip unchanged elapsed=\(elapsedMS(since: start))ms")
                return
            }
            DiagLog.log("sync", "autoPush write begin encryptedBytes=\(data.count) plainBytes=\(plain.count)")
            try sync.push(data)
            sync.lastHash = hash
            DiagLog.log("sync", "autoPush ok elapsed=\(elapsedMS(since: start))ms")
        } catch {
            DiagLog.log("sync", "autoPush failed elapsed=\(elapsedMS(since: start))ms error=\(error.localizedDescription)")
        }
    }

    /// Pull the folder's bundle and merge it into this device (every category).
    /// No-ops quietly when auto-sync isn't set up or there's nothing to pull.
    static func autoPull(sync: SyncStore, hosts: HostStore, knownHosts: KnownHostsStore, agent: AgentStore, snippets: SnippetStore) async {
        let start = Date()
        DiagLog.log("sync", "autoPull start configured=\(sync.isConfigured) passphrase=\(sync.hasPassphrase)")
        guard sync.autoEnabled else {
            DiagLog.log("sync", "autoPull skip disabled")
            return
        }
        guard let pass = sync.loadPassphrase() else {
            DiagLog.log("sync", "autoPull skip no passphrase")
            return
        }
        do {
            let data = try sync.pull()
            DiagLog.log("sync", "autoPull read bytes=\(data.count)")
            let bundle = try ConfigService.decode(data, passphrase: pass)
            // Snippets come across too (synced ones land with run-on-connect off in apply()).
            let selection = ConfigService.ImportSelection(hosts: true, settings: true, knownHosts: true,
                                                          keys: true, securityKeys: true, snippets: true)
            await ConfigService.apply(bundle, selection: selection, hosts: hosts, knownHosts: knownHosts, agent: agent, snippets: snippets)
            DiagLog.log("sync", "autoPull ok elapsed=\(elapsedMS(since: start))ms")
        } catch {
            DiagLog.log("sync", "autoPull failed elapsed=\(elapsedMS(since: start))ms error=\(error.localizedDescription)")
        }
    }

    private static func elapsedMS(since start: Date) -> Int {
        Int(Date().timeIntervalSince(start) * 1000)
    }
}

File: android/app/src/main/kotlin/cc/bsns/ssh/ConfigSync.kt
package cc.bsns.ssh

import android.content.Context
import android.net.Uri
import android.security.keystore.KeyGenParameterSpec
import android.security.keystore.KeyProperties
import androidx.documentfile.provider.DocumentFile
import cc.bsns.ssh.core.ConfigEnvelope
import java.security.KeyStore
import java.security.MessageDigest
import java.util.Base64
import javax.crypto.Cipher
import javax.crypto.KeyGenerator
import javax.crypto.SecretKey
import javax.crypto.spec.GCMParameterSpec

/**
 * Auto-sync settings: a user-chosen folder (any Files provider — iCloud Drive,
 * Drive, Dropbox, local) and a passphrase, both under the user's control. The
 * passphrase is wrapped by a non-extractable Keystore key and never leaves the
 * device in the clear; the provider only ever sees the AES-GCM ciphertext bundle.
 */
class SyncStore(context: Context) {
    private val prefs = context.getSharedPreferences("sync", Context.MODE_PRIVATE)
    private val ks = KeyStore.getInstance("AndroidKeyStore").apply { load(null) }
    private val wrapAlias = "bsns-sync-wrap"

    private fun wrapKey(): SecretKey {
        (ks.getEntry(wrapAlias, null) as? KeyStore.SecretKeyEntry)?.let { return it.secretKey }
        val kg = KeyGenerator.getInstance(KeyProperties.KEY_ALGORITHM_AES, "AndroidKeyStore")
        kg.init(KeyGenParameterSpec.Builder(wrapAlias, KeyProperties.PURPOSE_ENCRYPT or KeyProperties.PURPOSE_DECRYPT)
            .setBlockModes(KeyProperties.BLOCK_MODE_GCM)
            .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_NONE).build())
        return kg.generateKey()
    }

    val enabled: Boolean get() = folderUri != null && passphrase != null

    var folderUri: String?
        get() = prefs.getString("folder", null)
        private set(v) { prefs.edit().putString("folder", v).apply() }

    var passphrase: String?
        get() = prefs.getString("pass", null)?.let { stored ->
            runCatching {
                val blob = Base64.getDecoder().decode(stored)
                Cipher.getInstance("AES/GCM/NoPadding")
                    .apply { init(Cipher.DECRYPT_MODE, wrapKey(), GCMParameterSpec(128, blob.copyOfRange(0, 12))) }
                    .doFinal(blob.copyOfRange(12, blob.size)).toString(Charsets.UTF_8)
            }.getOrNull()
        }
        private set(v) {
            if (v == null) { prefs.edit().remove("pass").apply(); return }
            val cipher = Cipher.getInstance("AES/GCM/NoPadding").apply { init(Cipher.ENCRYPT_MODE, wrapKey()) }
            val blob = cipher.iv + cipher.doFinal(v.toByteArray(Charsets.UTF_8))
            prefs.edit().putString("pass", Base64.getEncoder().encodeToString(blob)).apply()
        }

    /** Hash of the last bundle we pushed/pulled, so we skip redundant writes. */
    var lastHash: String?
        get() = prefs.getString("hash", null)
        set(v) { prefs.edit().putString("hash", v).apply() }

    fun configure(folder: String, pass: String) { folderUri = folder; passphrase = pass; lastHash = null }
    fun clear() { prefs.edit().clear().apply() }
}

/**
 * The auto-sync engine. Pushes an encrypted bundle to the chosen folder and
 * pulls + merges it back. Additive merge (never deletes), so a synced device
 * gains the union — matches the manual import semantics. Same on-disk format as
 * the manual backup, so the two interoperate.
 */
object ConfigSync {
    // Shared cross-platform sync filename (matches iOS SyncStore.fileName), so a
    // folder synced between an iOS and an Android device points at one file.
    private const val FILE = "bsns-ssh-sync.json"
    // The original Android-only name; still read so a device that synced before
    // the rename keeps finding its bundle.
    private const val LEGACY_FILE = "bsns-config-aesgcm-v1.json"

    private fun sha256(b: ByteArray): String =
        Base64.getEncoder().encodeToString(MessageDigest.getInstance("SHA-256").digest(b))

    private fun folder(context: Context, store: SyncStore): DocumentFile? =
        store.folderUri?.let { DocumentFile.fromTreeUri(context, Uri.parse(it)) }?.takeIf { it.canWrite() }

    /** Write the current config (incl. software keys) to the sync folder, encrypted.
     *  Skips the write if nothing changed since the last sync. Returns a status string. */
    fun push(context: Context): String {
        val store = SyncStore(context)
        val pass = store.passphrase ?: return "sync not set up"
        val dir = folder(context, store) ?: return "sync folder unavailable"
        val plain = ConfigBundle.build(context, includeKeys = true)
        val hash = sha256(plain)
        if (hash == store.lastHash) return "already up to date"
        val sealed = ConfigEnvelope.encrypt(plain, pass)
        // Always write the canonical (cross-platform) name so iOS finds it, even
        // if a legacy-named file is still present from before the rename.
        val file = dir.findFile(FILE) ?: dir.createFile("application/json", FILE) ?: return "couldn't write to the folder"
        context.contentResolver.openOutputStream(file.uri, "wt")?.use { it.write(sealed) } ?: return "write failed"
        store.lastHash = hash
        return "synced up"
    }

    /** Read + merge the folder's bundle into this device (additive). Returns the
     *  applied summary, or null if there's nothing to pull / it's unchanged. */
    fun pull(context: Context): ConfigBundle.Applied? {
        val store = SyncStore(context)
        val pass = store.passphrase ?: return null
        val dir = folder(context, store) ?: return null
        val file = (dir.findFile(FILE) ?: dir.findFile(LEGACY_FILE))?.takeIf { it.isFile } ?: return null
        val sealed = context.contentResolver.openInputStream(file.uri)?.use { it.readBytes() } ?: return null
        val plain = runCatching { ConfigEnvelope.decrypt(sealed, pass) }.getOrNull() ?: return null
        val o = ConfigBundle.parse(plain)
        // The user configured this folder + passphrase, so it's a trusted source:
        // merge every category. Snippets come across too, but apply() forces their
        // runOnConnect off — synced snippets never silently auto-execute.
        val sel = ConfigBundle.Selection(hosts = true, knownHosts = true, settings = true,
                                         keys = true, snippets = true)
        return ConfigBundle.apply(context, o, sel)
    }
}
